android/src/main/java/com/tailscale/ipn/ui/notifier/Notifier.kt

```diff
@@ -47,95 +41,99 @@ class Notifier(private val scope: CoroutineScope) {
             isReady.complete(true)
             Log.d("Notifier", "Notifier is ready")
         }
+    }
 
-        private fun generateSessionId(): String {
-            synchronized(sessionIdLock) {
-                sessionId += 1
-                return sessionId.toString()
-            }
+    // NotifyWatchOpt is a bitmask of options supplied to the notifier to specify which
+    // what we want to see on the Noitfy bus
+    private enum class NotifyWatchOpt(val value: Int) {
+        engineUpdates(1), initialState(2), prefs(4), netmap(8), noPrivateKey(16), initialTailFSShares(
+            32
+        )
+    }
+
+    private var watchers = HashMap<String, (Notify) -> Unit>()
+
+    init {
+        Log.d("Notifier", "Notifier created")
+    }
+
+    val state: StateFlow<Ipn.State> by lazy {
+        flow(Ipn.State.NoState) { it.State?.let { Ipn.State.fromInt(it) } }
+    }
+    val netmap: StateFlow<Netmap.NetworkMap?> by lazy {
+        flow(null) { it.NetMap }
+    }
+    val prefs: StateFlow<Ipn.Prefs?> by lazy {
+        flow(null) { it.Prefs }
+    }
+    val engineStatus: StateFlow<Ipn.EngineStatus?> by lazy {
+        flow(null) { it.Engine }
+    }
+    val tailFSShares: StateFlow<Map<String, String>?> by lazy {
+        flow(null) { it.TailFSShares }
+    }
+    val browseToURL: StateFlow<String?> by lazy {
+        flow(null) { it.BrowseToURL }
+    }
+    val loginFinished: StateFlow<String?> by lazy {
+        flow(null) { it.LoginFinished?.property }
+    }
+    val version: StateFlow<String?> by lazy {
+        flow(null) { it.Version }
+    }
+
+    private fun <T> flow(initial: T, extract: (Notify) -> T?): StateFlow<T> {
+        val flow = MutableStateFlow(initial)
+        watchIPNBus(NotifyWatchOpt.netmap.value or NotifyWatchOpt.prefs.value or NotifyWatchOpt.initialState.value) { notify ->
+            extract(notify)?.let { flow.value = it }
         }
+        return flow
     }
 
-    // Starts an IPN Bus watcher.  **This is blocking** and will not return until
-    // the watcher is stopped and must be executed in a suitable coroutine scope such
-    // as Dispatchers.IO
+    // Starts an IPN Bus watcher. This is non-blocking.
     private external fun startIPNBusWatcher(sessionId: String, mask: Int)
 
-    // Stops an IPN Bus watcher
+    // Stops an IPN Bus watcher. This is non-blocking.
     private external fun stopIPNBusWatcher(sessionId: String)
 
-    private var watchers = HashMap<String, Watcher>()
-
     // Callback from jni when a new notification is received
-    fun onNotify(notification: String, sessionId: String) {
-        val notify = decoder.decodeFromString<Notify>(notification)
+    @OptIn(ExperimentalSerializationApi::class)
+    fun onNotify(notification: ByteArray, sessionId: String) {
+        val notify = decoder.decodeFromStream<Notify>(notification.inputStream())
```

Comment:
Re OOM, I'm wondering if the large allocations on corp are what's killing us.   In go land, this is a Reader, but we're just reading the entire works, then allocating mountains of strings to parse everything, etc.   Chunking onNotify might help with the large allocations in JNI - which is what seems to be failing.